tc/core/tensor.h

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <dlpack/dlpack.h>
+
+#include "tc/proto/compcache.pb.h"
+
+/**
+ * Various tensor utilities used in the Tensor Comprehensions compiler.
+ * At a high-level we use the DLTensor type from dlpack and we add a
+ * DLConstTensor to propagate const-correctness.
+ * DLTensor is used as an ML-framework and backend-agnostic representation to
+ * interact with user-level tensors.
+ * DLTensorUPtr and DLConstTensorUPtr are used as metadata-owning pointers.
+ * The underlying concrete device pointer is borrowed when calling
+ * ExecutionEngine::compile and ExecutionEngine::run.
+ *
+ * Internally, TC uses TensorInfo for compiling and interacting with the
+ * compilation caches. TensorInfo is backed by protobuf and stored directly in
+ * the caches.
+ */
+
+/**
+ * DLPack is missing a ConstTensor. Add it here for now.
+ * In the future, this would be contributed back to dlpack/include/dlpack.h.
+ */
+typedef struct {
+  const void* data;
+  DLContext ctx;
+  int ndim;
+  DLDataType dtype;
+  int64_t* shape;
+  int64_t* strides;
+  uint64_t byte_offset;
+} DLConstTensor;
+
+namespace tc {
+/**
+ * C++ unique_ptr abstraction on top of DLTensor/DLConstTensor
+ */
+struct DLTensorDeleter {
+  template <typename DLTensorType>
+  void operator()(const DLTensorType* t);
+};
+using DLTensorUPtr = std::unique_ptr<DLTensor, DLTensorDeleter>;
+using DLConstTensorUPtr = std::unique_ptr<DLConstTensor, DLTensorDeleter>;
+
+// Return non-owning raw pointers to DLTensor from metadata-owning DLTensorUPtr
+std::vector<const DLConstTensor*> extractRawPtrs(
+    const std::vector<DLConstTensorUPtr>& uptrs);
+std::vector<const DLTensor*> extractRawPtrs(
+    const std::vector<DLTensorUPtr>& uptrs);
+
+/**
+ * TensorInfo wraps the necessary tensor information to compile TCs and
+ * interact with the various caches.
+ * Notably, it contains alignment information but no underlying data pointer.
+ * It is serializable to protobuf and stored directly in the cache.
+ */
+struct TensorInfo {
+  DLDataType dtype;
+  uint64_t alignment;
```

Comment:
how do you use this alignment if you don't control the pointer allocation?

Reply:
we are not using it yet; in the future it will be used to switch between different vectorized versions of a same kernel